### ISSUE TITLE
borrow_deref_ref false positive on closure

### ISSUE THREAD

**Issue by stormshield-gt** [rust-lang/rust-clippy#16556]
### Summary

The `borrow_deref_ref` is triggered even if it's sometimes necessary inside closure.

There is discussion about why it compiles [here](https://users.rust-lang.org/t/closure-that-borrow-from-their-captures/138226) 

### Lint Name

borrow_deref_ref

### Reproducer

I tried this code:

```rust
use std::pin::Pin;

#[tokio::main]
async fn main() {

    // Doesn't compile.
    // for_each_city_box(|city| Box::pin(async { println!("{city}") })).await;

    // But with reborrowing it does.
    for_each_city(|city| {
        Box::pin(async {
            let city = &*city;
            println!("{city}")
        })
    })
    .await;
}

async fn for_each_city<F>(mut f: F)
where
    F: for<'c> FnMut(&'c str) -> Pin<Box<dyn Future<Output = ()> + Send + 'c>>,
{
    for x in ["New York", "London", "Tokyo"] {
        f(x).await;
    }
}
```

I saw this happen:

```
$ cargo clippy
warning: deref on an immutable reference
  --> src/main.rs:12:24
   |
12 |             let city = &*city;
   |                        ^^^^^^ help: if you would like to reborrow, try removing `&*`: `city`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#borrow_deref_ref
   = note: `#[warn(clippy::borrow_deref_ref)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground" -p playground` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
```

I expected to see this happen:

Don't have any warning

### Version

```text
clippy 0.1.93 (254b59607d 2026-01-19)
```

**Comment by Jarcho** [comment]
In this case you could also use either `async move` or `println!("{}", *city)`.

In general this lint should be ignoring `let x = &*upvar` the alternative can, in some cases, be quite annoying to write.

**Comment by stormshield-gt** [comment]
I confirm that, in my real world codebase, I don't want to use `async move` because I have to create several `async` blocks that need to capture the same data.
I could use `clone` everywhere + `async move` but I think it’s better to just reborrow the data.

By the way, here it's a more minimalist reproducer

```rust
pub fn f(s: &String) -> impl FnOnce() {
    || {
        &*s;
    }
}
```
